lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -527,3 +527,47 @@ impl<T> TakePointer<*mut T> for *mut T {
 		ret
 	}
 }
+
+
+pub(crate) mod ObjOps {
+	#[inline]
+	#[must_use = "returns new dangling pointer"]
+	pub(crate) fn heap_alloc<T>(obj: T) -> *mut T {
+		let ptr = Box::into_raw(Box::new(obj));
+		nonnull_ptr_to_inner(ptr)
+	}
+	#[inline]
+	pub(crate) fn nonnull_ptr_to_inner<T>(ptr: *const T) -> *mut T {
+		if core::mem::size_of::<T>() == 0 {
+			// We map `None::<T>` as `T { inner: null, .. }` which works great for all non-ZST `T`.
+			// For ZSTs, we need to differentiate between null implying `None` and null implying
+			// `Some` with no allocation.
+			// Thus, for ZSTs, we add one (usually) page here, which should always be aligned.
+			// Note that this relies on undefined behavior! A pointer to NULL may be valid, but a
+			// pointer to NULL + 4096 is almost certainly not. That said, Rust's existing use of
+			// `(*mut T)1` for the pointer we're adding to is also not defined, so we should be
+			// fine.
+			debug_assert_eq!(ptr as usize, 1);
+			unsafe { (ptr as *mut T).cast::<u8>().add(4096 - 1).cast::<T>() }
```

Comment:
and a comment explaining why for zero-sized types, we add 4095 and for non-zero-sized types, 4096.